src/server/service/kv.rs

```diff
@@ -1392,6 +1392,11 @@ fn future_scan<E: Engine, L: LockMgr>(
             if let Some(err) = extract_region_error(&v) {
                 resp.set_region_error(err);
             } else {
+                resp.set_has_locks(
```

Comment:
This may introduce the cost scanning the result vector another time which the `tispark` client wants to avoid. 

The current `tispark` scan process is like this:
- 1.Send the `ScanRequest` to tikv.
- 2.Get the kv pairs in the `ScanResponse`.
- 3.Loop through all the kv pairs, collect the locks if there is.
- 4.Batch resolve all the encountered locks, that is send `BatchResolveRequest` to tikv.
- 5.Retry the `ScanRequest` until there is no lock error.

The step 3 may introduce much cost for every scan request which hurts the performance a lot. So we may need to collect all the locks in the response, so if the `len(locks) > 0` the tispark could try to batch resolve these locks directly using the information from the `ScanResponse` and step 3 could be avoided.

If the `tispark` server supports triggering resolve the locks for a whole region like `tidb`, we could break the tikv scanner if there is a lock encountered directly and return the lock error in the response. But seems tispark only supports batch resolve by now.

Reply:
> This may introduce the cost scanning the result vector another time which the `tispark` client wants to avoid. 
> 
> The current `tispark` scan process is like this:
> - 1.Send the `ScanRequest` to tikv.
> - 2.Get the kv pairs in the `ScanResponse`.
> - 3.Loop through all the kv pairs, collect the locks if there is.
> - 4.Batch resolve all the encountered locks, that is send `BatchResolveRequest` to tikv.
> - 5.Retry the `ScanRequest` until there is no lock error.
> 
> The step 3 may introduce much cost for every scan request which hurts the performance a lot. So we may need to collect all the locks in the response, so if the `len(locks) > 0` the tispark could try to batch resolve these locks directly using the information from the `ScanResponse` and step 3 could be avoided.
> 
> If the `tispark` server supports triggering resolve the locks for a whole region like `tidb`, we could break the tikv scanner if there is a lock encountered directly and return the lock error in the response. But seems tispark only supports batch resolve by now.

thx for your explanation. I believe it is a right way not to break the tikv scanner here. It will not break the resolve lock logic for elder versions of tidb and tispark.

On the other hand, while returning all locks in response, it's still possible to implement the tispark logic to trigger a region-wise lock resolver if necessary.

Reply:
> This may introduce the cost scanning the result vector another time which the tispark client wants to avoid.

Returning the first lock from the scanner is also O(N), I think it won't make much difference?